### ISSUE TITLE
Implement proper item serialization

### ISSUE THREAD

**Issue by collinsmith** [collinsmith/riiablo#93]
Item serialization has been placeholder and needs to be properly implemented. Currently item serialization data is just a reference to the data as it was deserialized (i.e., as it was within the d2s). This issue will require implementing support within `BitStream` for writing values. Ideally the deserialized format should match the original serialized format to maintain compatibility.

**Comment by collinsmith** [comment]
I've been researching and playing around with various APIs regarding support for this issue. I've decided to settle on creating `com.riiablo.io` package since this spec is growing and have separate classes for input and output. Reading and writing in the same stream should not be required by Riiablo, and doing it this way seems easier.

Looking into `bit-io` library for strategy -- some good ideas, but not completely sufficient for my needs (slicing buffer). I may also couple the I/O package with netty `ByteBuf`, since it supports pooling and is compatible with network I/O (eventually I'd like to have all codecs use `ByteBuf`, pending research).

Long term, if it's not too expensive, I'd like to look into using this for all I/O. This would mean that all codecs use the same code base, and more pertinently, when a codec needs to create a slice for a section, it can do so easily. The bit input API will also support recovery at a specified signature -- which will be helpful if the codec can recover and log a warning instead of a fatal error. I'm also going to guarantee that byte position can be retrieved, so tracking down errors will report offset within file (this gets a little messy when inside something like an item bit stream, so offset would probably be the start of item's input slice).

**Comment by collinsmith** [comment]
dd1444ad9a143fe4f67debd1df828615e7847115 created `com.riiablo.io` package. I'm happy with this API so far -- supports seamless reading from bit stream (aligned and unaligned), enforcing alignment only when using no-args read methods. The idea is that bit stream read operations should use methods that specify bits, while byte stream operations should use no-args methods. This may cause issues down the line if some bit stream read operations use no-args methods when they just happen to be aligned. Alternatively, I could modify the API to work as follows:

```
ByteInput align() {...}
align().read8u();
BitInput unalign() {...}
unalign().read7u(7);
```

**Comment by collinsmith** [comment]
Basic serialization is complete. Going to tighten up code where I can and push.

**Comment by collinsmith** [comment]
Serialization tested on a large variety of items, so far so good. I'm going to start working on deprecating/removing cache variables such as `Item#data` so I can close this issue.

**Comment by collinsmith** [comment]
Fixed by 2beecb35a8f7ef1d9b5c43c368f63c9dd51bef3f. `Item#data` issue will be addressed by #98
